### PR TITLE
feat(app) add k8s event for Application

### LINKED ISSUE #1424: [Feature] Add k8s event for Application Object on every step
We should add events for every step of the application object state in Application controller.

Currently, the events is null when we use `kubectl describe` to see the application object.

```
$ kubectl describe application app-basic
Name:         app-basic
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  core.oam.dev/v1beta1
Kind:         Application
Metadata:
  Creation Timestamp:  2021-03-31T12:57:53Z
  Generation:          1
Spec:
  Components:
    Name:  app-basic
    Properties:
      Image:  nginx:1.9.4
      Port:   8080
    Scopes:
      healthscopes.core.oam.dev:  app-basic-default-health
    Traits:
      Properties:
        Replicas:  2
      Type:        scaler
    Type:          webservice
Status:
  Batch Rolling State:
  Components:
    API Version:  core.oam.dev/v1alpha2
    Kind:         Component
    Name:         app-basic

  Conditions:
    Last Transition Time:  2021-03-31T12:57:53Z
    Reason:                Available
    Status:                True
    Type:                  Parsed
    Last Transition Time:  2021-03-31T12:57:53Z
    Reason:                Available
    Status:                True
    Type:                  Built
    Last Transition Time:  2021-03-31T12:57:55Z
    Reason:                Available
    Status:                True
    Type:                  Applied
    Last Transition Time:  2021-03-31T12:57:55Z
    Reason:                Available
    Status:                True
    Type:                  HealthCheck
  Current Batch:           0
  Latest Revision:
    Name:           app-basic-v1
    Revision:       1
    Revision Hash:  d22af8f24d93fd8b
  Rolling State:
  Services:
    Healthy:  true
    Name:     app-basic
    Traits:
      Healthy:              true
      Type:                 scaler
  Status:                   running
  Upgraded Ready Replicas:  0
  Upgraded Replicas:        0
Events:                    null
```

We should record all important state into events, such as `parsed successfully`, `rendered successfully`, `applied successfuly`, `health checked healthy`

### PR DESCRIPTION
**Pull request by Ghostbaby**
add k8s event for Application

fixed #1424

Signed-off-by: zhuhuijun <zhuhuijunzhj@gmail.com>

### CODE REVIEW

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go [hunk 1]
```diff
@@ -180,6 +193,7 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	}
 	app.Status.Services = appCompStatus
 	app.Status.SetConditions(readyCondition("HealthCheck"))
+	r.Recorder.Event(app, corev1.EventTypeNormal, ReasonHealthCheck, MessageHealthCheck)
 	app.Status.Phase = common.ApplicationRunning
 	err = handler.garbageCollection(ctx)
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:198]: please also add a final event after the GC

File: pkg/controller/core.oam.dev/v1alpha2/application/event.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	MessageParsed      = "Parsed successfully"
+	MessageRendered    = "Rendered successfully"
+	MessageApplied     = "Applied successfully"
+	MessageHealthCheck = "Health checked healthy"
+
+	MessageFailedParse       = "fail to parse application, err: %v"
+	MessageFailedRender      = "fail to render application, err: %v"
+	MessageFailedApply       = "fail to apply component, err: %v"
+	MessageFailedHealthCheck = "fail to health check, err: %v"
+)
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/event.go:43]: Let's merge these types into `apis/types`

File: pkg/controller/core.oam.dev/v1alpha2/application/suite_test.go [hunk 3]
```diff
@@ -122,11 +124,12 @@ var _ = BeforeSuite(func(done Done) {
...
-		Log:    ctrl.Log.WithName("Application-Test"),
-		Scheme: testScheme,
-		dm:     dm,
-		pd:     pd,
+		Client:   k8sClient,
+		Log:      ctrl.Log.WithName("Application-Test"),
+		Scheme:   testScheme,
+		dm:       dm,
+		pd:       pd,
+		Recorder: record.NewFakeRecorder(10),
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/suite_test.go:132]: what's the `10` mean? Will it affect our CI?

**Review by wonderflow — CHANGES_REQUESTED** [review]
Please add a test to check the event added and fix CI.

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go (hunk 1 above)
Comment (Ghostbaby) [reply pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:198]: ok

File: pkg/controller/core.oam.dev/v1alpha2/application/event.go (hunk 2 above)
Comment (Ghostbaby) [reply pkg/controller/core.oam.dev/v1alpha2/application/event.go:43]: ok

File: pkg/controller/core.oam.dev/v1alpha2/application/suite_test.go (hunk 3 above)
Comment (Ghostbaby) [reply pkg/controller/core.oam.dev/v1alpha2/application/suite_test.go:132]: this is event channel size, maybe it affects the ci, I will set a larger num.

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go [hunk 4]
```diff
@@ -331,6 +331,30 @@ spec:
...
+		appKey := client.ObjectKey{
+			Name:      appwithNoTrait.Name,
+			Namespace: appwithNoTrait.Namespace,
+		}
+		reconcileRetry(reconciler, reconcile.Request{NamespacedName: appKey})
+
+		events := collectEvents(recorder.Events)
+		Expect(len(events)).ShouldNot(Equal(0))
+		for _, event := range events {
+			Expect(event).Should(ContainSubstring(corev1.EventTypeNormal))
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go:353]: Will other negtive case affect this test result? If I understand correct, this events will contain all events from reconcile in this suit test. 

We'd better compare more detailed events info, and narrow the scope.

File: pkg/controller/core.oam.dev/v1alpha2/application/suite_test.go [hunk 5]
```diff
@@ -55,6 +57,7 @@ import (
 // These tests use Ginkgo (BDD-style Go testing framework). Refer to
 // http://onsi.github.io/ginkgo/ to learn more about Ginkgo.
 var cfg *rest.Config
+var recorder = record.NewFakeRecorder(10000)
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/suite_test.go:60]: I noticed the fakeRecorder is very simple, it doesn't include the appname in the event message, so you can hardly distinguish them . I suggest you can re-write that package here. At least include the object name.

```

func (f *FakeRecorder) Event(object runtime.Object, eventtype, reason, message string) {
	if f.Events != nil {
		f.Events <- fmt.Sprintf("%s %s %s %s", object.Name(),eventtype, reason, message)
	}
}

```

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go (hunk 4 above)
Comment (Ghostbaby) [reply pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go:353]: copy

File: apis/types/event.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+	ReasonFailedHealthCheck = "FailedHealthCheck"
+)
+
+// event message for Application
+const (
+	MessageParsed      = "Parsed successfully"
+	MessageRendered    = "Rendered successfully"
+	MessageApplied     = "Applied successfully"
+	MessageHealthCheck = "Health checked healthy"
+	MessageReconciled  = "Reconciled successfully"
```
Comment (wonderflow) [inline apis/types/event.go:39]:
```suggestion
	MessageReconciled  = "Deployed successfully"
```

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go [hunk 7]
```diff
@@ -1523,6 +1604,16 @@ func reconcileRetry(r reconcile.Reconciler, req reconcile.Request) {
 	}, 30*time.Second, time.Second).Should(BeNil())
 }
 
+func reconcileOnce(r reconcile.Reconciler, req reconcile.Request) {
+	Eventually(func() error {
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go:1608]: if we just reconcile once, we don't need `Eventually` func

File: pkg/controller/core.oam.dev/v1alpha2/application/suite_test.go [hunk 8]
```diff
@@ -166,3 +171,49 @@ var _ = AfterSuite(func() {
 	Expect(err).ToNot(HaveOccurred())
 	close(stop)
 })
+
+func collectEvents(source <-chan string) []string {
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/suite_test.go:175]: I think we can even make the function as a method of `FakeRecorder`. And the fake recorder should record all events receved.

```
type FakeRecorder struct {
	Events chan string
        Message map[string][]string
}
```

We can filter all event messages for separated app name in message map.

And and a functon to get events by specified app.

**Comment by zzxwill** [comment]
@Ghostbaby does `feat` in the tile mean `feature`?

**Comment by Ghostbaby** [comment]
> @Ghostbaby does `feat` in the tile mean `feature`?

yes

File: pkg/controller/core.oam.dev/v1alpha2/application/suite_test.go (hunk 8 above)
Comment (Ghostbaby) [reply pkg/controller/core.oam.dev/v1alpha2/application/suite_test.go:175]: copy

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go [hunk 9]
```diff
@@ -181,6 +196,7 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	}
 	app.Status.Services = appCompStatus
 	app.Status.SetConditions(readyCondition("HealthCheck"))
+	r.Recorder.Event(app, corev1.EventTypeNormal, velatypes.ReasonHealthCheck, velatypes.MessageHealthCheck)
 	app.Status.Phase = common.ApplicationRunning
 	err = garbageCollection(ctx, handler)
 	if err != nil {
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:202]: please also add an event when GC has error

File: pkg/controller/core.oam.dev/v1alpha2/application/suite_test.go [hunk 10]
```diff
@@ -168,3 +175,72 @@ var _ = AfterSuite(func() {
...
+		}
+
+		event := &Events{
+			Name:      objectMeta.GetName(),
+			Namespace: objectMeta.GetNamespace(),
+			EventType: eventtype,
+			Reason:    reason,
+			Message:   message,
+		}
+		f.Events <- fmt.Sprintf("%s %s %s %s %s", objectMeta.GetName(), objectMeta.GetNamespace(), eventtype, reason, message)
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/suite_test.go:206]: It's not necessary as you have stored all the information in `f.Message`.

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go (hunk 9 above)
Comment (Ghostbaby) [reply pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:202]: copy

File: pkg/controller/core.oam.dev/v1alpha2/application/suite_test.go (hunk 10 above)
Comment (Ghostbaby) [reply pkg/controller/core.oam.dev/v1alpha2/application/suite_test.go:206]: copy

**Review by wonderflow — APPROVED** [review]

**Pull request merged**